EKF/control.cpp

```diff
@@ -757,10 +757,19 @@ void Ekf::controlHeightFusion()
 		// we have just switched to using range finder, calculate height sensor offset such that current
 		// measurment matches our current height estimate
 		if (_control_status_prev.flags.rng_hgt != _control_status.flags.rng_hgt) {
-			if (_terrain_initialised) {
-				_hgt_sensor_offset = _terrain_vpos;
+			// use the parameter rng_gnd_clearance if on ground to avoid a noisy offset initialization (e.g. sonar)
+			if (_control_status.flags.in_air) {
+
+				if (_terrain_initialised) {
+
+					_hgt_sensor_offset = _terrain_vpos;
+				} else {
+
+					_hgt_sensor_offset = _R_rng_to_earth_2_2 * _range_sample_delayed.rng + _state.pos(2);
```

Comment:
> If you switch to range finder, the estimate will be the current height.

I beg to differ, but please correct me if I am wrong. This line https://github.com/PX4/ecl/blob/8ed61ac0520686041e3ee193e693639d97b663db/EKF/vel_pos_fusion.cpp#L162 says `z_innovation = z_estimated - (-range + terrain_vpos)`
In other words z_estimated is corrected towards `(-range + terrain_vpos)`, which is not the height, but the height + estimated terrain altitude when range sensor was activated.